tensorboard/manager.py

```diff
@@ -16,34 +16,44 @@
 
 
 import base64
-import collections
+import dataclasses
 import datetime
 import errno
 import json
 import os
 import subprocess
 import tempfile
 import time
+import typing
 
+from typing import Callable, Optional, Type, Union
 
 from tensorboard import version
 from tensorboard.util import tb_logging
 
 
-# Type descriptors for `TensorBoardInfo` fields.
-#
-# We represent timestamps as int-seconds-since-epoch rather than
-# datetime objects to work around a bug in Python on Windows. See:
-# https://github.com/tensorflow/tensorboard/issues/2017.
-_FieldType = collections.namedtuple(
-    "_FieldType",
-    (
-        "serialized_type",
-        "runtime_type",
-        "serialize",
-        "deserialize",
-    ),
-)
+@dataclasses.dataclass(frozen=True)
+class _FieldType:
```

Comment:
So - now that we can use dataclasses, I don't think FieldType really makes sense any more. With namedtuple, there wasn't a great way to specify the type of each field, hence the separate OrderedDict, but the dataclass solves that since we can use the type annotations (as you have it now).

However, when we're using type annotations, it's weird to say that the type of the field is a _FieldType instance - the actual type of the `version` field on a real TensorBoardInfo instance is just `str`, it's not `_type_str` (which is actually a dataclass object of type _FieldType).  Instead, I think it would be more natural to just replace `_type_str` with `str` and `_type_int` with `int` in the annotations of TensorBoardInfo.

In theory, FieldType allowed us to support types beyond `int` and `str` that might have required custom serialization and deserialization functions, but in practice, it's been several years and we are still only using `int` and `str`.  So I don't think the complexity is really worth it - we can just update the functions to special case int and str and handle them appropriately. (And for the _info_to_string we could perhaps even just use `asdict()` and then the default json serialization.)

This might get a bit complicated for this PR, so my suggestion would be to pull out the manager.py change (and the test change) into its own PR to make it easier to review.

Reply:
Reverted the changes for now, will send a `part 2` PR later.